### PR TITLE
Allocation improvement in precalculating layouts

### PR DESCRIPTION
**Pull request by snakefoot**
Skip enumerator allocation. Maybe consider creating a struct-enumerator for the PropertiesDictionary.

I have dared to increase the allowed parameter count in defer-formatting-check from 3 to 5, after the optimization introduced with  #2595

### CODE REVIEW

**Comment by 304NotModified** [comment]
PS TestStructuredProperties_json_async is broken now on both CI

**Comment by snakefoot** [comment]
Yes I know. It is actually a different bug that has been revealed. Going to fix it when I get back from the UK



Sent from my Samsung device


-------- Original message --------

**Comment by snakefoot** [comment]
@304NotModified Fixed the bug that was uncovered. Will update the NLog Performance Test, so this allocation bug would be detected.

**Comment by snakefoot** [comment]
Some performance numbers:

| Logger Name      | Messages   | Size | Args | Threads | Loggers |
|------------------|------------|------|------|---------|---------|
| PerfLogger       | 15.000.000 |  128 |    2 |       1 |       1 |


| Test Name  | Time (ms) | Msgs/sec  | GC2 | GC1 | GC0 | CPU (ms) | Alloc (MB) |
|------------|-----------|-----------|-----|-----|-----|----------|------------|
| NLog 4.4   |    27.732 |   540.878 |   0 |  49 |  99 |   33.218 |    6.128,3 |
| Master    |    24.450 |   613.478 |   0 |  19 |  43 |   29.218 |    2.176,6 |
| This PR    |    24.396 |   614.842 |   0 |  21 |  43 |   29.234 |    2.176,6 |

| Logger Name      | Messages   | Size | Args | Threads | Loggers |
|------------------|------------|------|------|---------|---------|
| PerfLogger       | 15.000.000 |  128 |    6 |       1 |       1 |

| Test Name  | Time (ms) | Msgs/sec  | GC2 | GC1 | GC0 | CPU (ms) | Alloc (MB) |
|------------|-----------|-----------|-----|-----|-----|----------|------------|
| NLog 4.4    |    18.169 |   825.579 |   0 |  35 |  71 |   33.015 |    6.127,1 |
| Master    |    25.908 |   578.965 |   0 |  48 | 101 |   44.343 |    6.067,9 |
| This PR    |    18.950 |   791.521 |   0 |  52 | 113 |   23.156 |    6.067,7 |

You can see that `master` is currently is doing double formatting when Args = 6. This is now fixed with this PR.

**Comment by 304NotModified** [comment]
great results!

File: src/NLog/LogEventInfo.cs [hunk 1]
```diff
@@ -593,33 +593,89 @@ private static bool NeedToPreformatMessage(object[] parameters)
...
+
+                    // Skip enumerator allocation when all properties comes from the message-template
+                    for (int i = 0; i < properties.MessageProperties.Count; ++i)
+                    {
+                        var property = properties.MessageProperties[i];
+                        if (IsSafeToDeferFormatting(property.Value))
+                        {
+                            immutableProperties = false;
+                            break;
+                        }
```
[the remark this answers is not in the document]
Comment (304NotModified) [reply src/NLog/LogEventInfo.cs]: I meant 637 - 644

[the remark this answers is not in the document]
Comment (snakefoot) [reply src/NLog/LogEventInfo.cs]: And what are the line numbers for "below" ? Or did my initial comment explain why?

[the remark this answers is not in the document]
Comment (304NotModified) [reply src/NLog/LogEventInfo.cs]: > Have you looked at the comments around the two code-blocks?

Yes, but I'm afraid I cannot explain it better for now. It's also not that important, we could refactor it in after merge of the PR.

[the remark this answers is not in the document]
Comment (snakefoot) [reply src/NLog/LogEventInfo.cs]: I have translated your comment into extracting the logic from the two sections into a separate method called `HasImmutableProperties`.

**Comment by snakefoot** [comment]
Now also discovered that the #2653 did not make proper use of  #2653. This has now been fixed. Improves concurrency when many threads are writing mutable LogEvents to the same target.

**Comment by snakefoot** [comment]
@304NotModified Think it is ready for review / merge now.

**Comment by 304NotModified** [comment]
Thanks!

**Pull request merged**